File: problems/arrays-and-hashing/6-group-anagrams/rust/src/main.rs
use std::collections::HashMap;

fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
    let mut res: HashMap<Vec<i32>, Vec<String>> = HashMap::new();

    for s in strs {
        let mut count: Vec<i32> = vec![0; 26];

        for c in s.chars() {
            count[c as usize - 'a' as usize] += 1;
        }

        res.entry(count).or_insert(Vec::new()).push(s);
    }

    res.into_values().collect()
}

fn main() {
    let strs_1 = vec![String::from("act"), String::from("pots"), String::from("tops"), String::from("cat"), String::from("stop"), String::from("hat")];
    let strs_2 = vec![String::from("x")];
    let strs_3 = vec![String::from("")];

    let mut output_1 = vec![vec![String::from("hat")], vec![String::from("act"), String::from("cat")], vec![String::from("stop"), String::from("pots"), String::from("tops")]];
    let mut output_2 = vec![vec![String::from("x")]];
    let mut output_3 = vec![vec![String::from("")]];
    output_1.iter_mut().for_each(|g| g.sort());
    output_2.iter_mut().for_each(|g| g.sort());
    output_3.iter_mut().for_each(|g| g.sort());
    output_1.sort();
    output_2.sort();
    output_3.sort();

    let mut res_1 = group_anagrams(strs_1);
    let mut res_2 = group_anagrams(strs_2);
    let mut res_3 = group_anagrams(strs_3);
    res_1.iter_mut().for_each(|g| g.sort());
    res_2.iter_mut().for_each(|g| g.sort());
    res_3.iter_mut().for_each(|g| g.sort());
    res_1.sort();
    res_2.sort();
    res_3.sort();

    if output_1 != res_1 || output_2 != res_2 || output_3 != res_3 {
        println!("Actual results not equal with expected results.");
    }
    else {
        println!("Results correct.");
    }
}

File: problems/arrays-and-hashing/6-group-anagrams/cpp/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

// HashMap method is complicated for C++
// Using the "String sorting" method
vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> groups;

    for (auto s : strs) {
        string sortedString = s;
        sort(sortedString.begin(), sortedString.end());
        if (!groups.contains(sortedString)) {
            groups[sortedString] = {};
        }
        auto& group = groups[sortedString];
        group.push_back(s);
    }

    vector<vector<string>> result;
    for (auto& [key, group] : groups)
        result.push_back(group);
    return result;
}

int main(int argc, char const *argv[])
{
    vector<string> strs1 = { "act", "pots", "tops", "cat", "stop", "hat" };
    vector<string> strs2 = { "x" };
    vector<string> strs3 = { "" };

    vector<vector<string>> output1 = { { "hat" }, { "act", "cat" }, { "stop", "pots", "tops" }};
    vector<vector<string>> output2 = { { "x" } };
    vector<vector<string>> output3 = { { "" } };
    for (auto& group : output1) sort(group.begin(), group.end());
    for (auto& group : output2) sort(group.begin(), group.end());
    for (auto& group : output3) sort(group.begin(), group.end());
    sort(output1.begin(), output1.end());
    sort(output2.begin(), output2.end());
    sort(output3.begin(), output3.end());

    auto res1 = groupAnagrams(strs1);
    auto res2 = groupAnagrams(strs2);
    auto res3 = groupAnagrams(strs3);
    for (auto& group : res1) sort(group.begin(), group.end());
    for (auto& group : res2) sort(group.begin(), group.end());
    for (auto& group : res3) sort(group.begin(), group.end());
    sort(res1.begin(), res1.end());
    sort(res2.begin(), res2.end());
    sort(res3.begin(), res3.end());

    if (res1 != output1 || res2 != output2 || res3 != output3) {
        cout << "Actual results not equal with expected results." << endl;
    }
    else {
        cout << "Results correct." << endl;
    }
    
    return 0;
}
